src/Analyzers/CSharp/Analyzers/UseParameterNullChecking/CSharpUseParameterNullCheckingDiagnosticAnalyzer.cs

```diff
@@ -0,0 +1,296 @@
+                    SyntaxKind.ConstructorDeclaration,
+                    SyntaxKind.MethodDeclaration,
+                    SyntaxKind.LocalFunctionStatement,
+                    SyntaxKind.SimpleLambdaExpression,
+                    SyntaxKind.ParenthesizedLambdaExpression,
+                    SyntaxKind.AnonymousMethodExpression,
+                    SyntaxKind.OperatorDeclaration,
+                    SyntaxKind.ConversionOperatorDeclaration);
+            });
+
+        private void AnalyzeSyntax(
+            SyntaxNodeAnalysisContext context,
+            IMethodSymbol argumentNullExceptionConstructor,
+            IMethodSymbol argumentNullExceptionStringConstructor,
+            IMethodSymbol? referenceEqualsMethod)
+        {
+            var cancellationToken = context.CancellationToken;
+
+            var semanticModel = context.SemanticModel;
+            var syntaxTree = semanticModel.SyntaxTree;
+
+            var option = context.Options.GetOption(CSharpCodeStyleOptions.PreferParameterNullChecking, syntaxTree, cancellationToken);
+            if (!option.Value)
+            {
+                return;
+            }
+
+            var node = context.Node;
+            var block = node switch
+            {
+                MethodDeclarationSyntax methodDecl => methodDecl.Body,
+                ConstructorDeclarationSyntax constructorDecl => constructorDecl.Body,
+                LocalFunctionStatementSyntax localFunctionStatement => localFunctionStatement.Body,
+                AnonymousFunctionExpressionSyntax anonymousFunction => anonymousFunction.Block,
+                OperatorDeclarationSyntax operatorDecl => operatorDecl.Body,
+                ConversionOperatorDeclarationSyntax conversionDecl => conversionDecl.Body,
+                _ => throw ExceptionUtilities.UnexpectedValue(node)
+            };
+
+            // More scenarios should be supported eventually: https://github.com/dotnet/roslyn/issues/58699
+            if (block is null)
+            {
+                return;
+            }
+
+            var methodSymbol = node is AnonymousFunctionExpressionSyntax
+                ? (IMethodSymbol?)semanticModel.GetSymbolInfo(node, cancellationToken).Symbol
+                : (IMethodSymbol?)semanticModel.GetDeclaredSymbol(node, cancellationToken);
+            if (methodSymbol is null || methodSymbol.Parameters.IsEmpty)
+            {
+                return;
+            }
+
+            foreach (var statement in block.Statements)
+            {
+                var parameter = TryGetParameterNullCheckedByStatement(statement);
+                if (parameter is not null
+                    && (!parameter.Type.IsValueType
+                        || parameter.Type.OriginalDefinition.SpecialType == SpecialType.System_Nullable_T
+                        || parameter.Type.TypeKind is TypeKind.Pointer or TypeKind.FunctionPointer)
```

Comment:
extract into helper plz (can be local function or method).  mixing the && || and ! is too much for my brain :)